### PR TITLE
feat(lib): implement executeOperationV2

### PR DESCRIPTION
**Pull request by kvwalker**
Fixes NODE-1896

### CODE REVIEW

File: lib/operation_classes/insert_one.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+class InsertOneOperation extends Operation {
+  constructor(collection, doc, options) {
+    super(options);
+
+    this.collection = collection;
+    this.doc = doc;
+    this.options = options;
+  }
+
+  static get aspects() {
```
[the remark this answers is not in the document]
Comment (daprahamian) [reply lib/operation_classes/insert_one.js]: Disagree on this one. We should have a constructor check in Base that asserts that `aspects` are defined for the class, else throw. There should never be an operation without an `aspects` property; it should instead be an empty set.

[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/operation_classes/insert_one.js]: This initial PR has "aspects" in it, but isn't concerned with supporting them (see the commented out check in `Operation`s ctor). I think ultimately no operation will have empty aspects, but I don't really see a need to enforce that there aren't any. I feel like seeing the boilerplate just for an empty set doesn't convey anything useful to me.

[the remark this answers is not in the document]
Comment (kvwalker) [reply lib/operation_classes/insert_one.js]: I think a good compromise would be to leave aspects out of this PR and make a ticket to support aspects in the future.

[the remark this answers is not in the document]
Comment (daprahamian) [reply lib/operation_classes/insert_one.js]: Sgtm

File: lib/operations/insert_one.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    const doc = this.doc;
+    const options = this.options;
+
+    if (Array.isArray(doc)) {
+      return callback(
+        MongoError.create({ message: 'doc parameter must be an object', driver: true })
+      );
+    }
+
+    insertDocuments(coll, [doc], options, (err, r) => {
```
Comment (daprahamian) [inline lib/operations/insert_one.js]: Note for future: Once we have more operations defined, I'm interested in how these common behaviors are shared between them.

File: lib/operations/execute_operation_v2.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * a single point for determining whether callbacks or promises should be used. Additionally
+ * it allows for a single point of entry to provide features such as implicit sessions, which
+ * are required by the Driver Sessions specification in the event that a ClientSession is
+ * not provided
+ *
+ * @param {object} topology The topology to execute this operation on
+ * @param {Operation} operation The operation to execute
+ * @param {function} callback The command result callback
+ */
+const executeOperationV2 = (topology, operation, callback) => {
```
Comment (mbroadst) [inline lib/operations/execute_operation_v2.js]: nit: is there benefit to this being a function expression vs just a function?

File: lib/operations/execute_operation_v2.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  if (topology == null) {
+    throw new TypeError('This method requires a valid topology instance');
+  }
+
```
Comment (mbroadst) [inline lib/operations/execute_operation_v2.js]: Let's make an explicit decision about whether we want to ensure that `operation instanceof OperationBase` or not. The benefit of not explicitly checking is that we can ducktype operations, the benefit of ensuring the check is that we can call `execute` without concern that it doesn't exist on the operation.

File: lib/operations/operation.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  get session() {
+    return this.options.session;
+  }
+
+  clearSession() {
+    delete this.options.session;
+  }
+
+  execute() {
+    throw new TypeError('`execute` must be implemented for Operation subclasses');
```
Comment (mbroadst) [inline lib/operations/operation.js]: `'OperationBase' subclasses`

File: lib/operations/operation.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+const Aspect = {
+  SKIP_SESSION: Symbol('SKIP_SESSION')
+};
+
+class OperationBase {
```
Comment (mbroadst) [inline lib/operations/operation.js]: can you add some basic documentation for this class please?

File: lib/operations/execute_operation_v2.js (hunk 4 above)
Comment (kvwalker) [reply lib/operations/execute_operation_v2.js]: I think we should ensure that `operation instanceof OperationBase` so I've made that change.

Comment (daprahamian) [reply lib/operations/execute_operation_v2.js]: Should we not still be checking that `topology != null`?

Comment (mbroadst) [reply lib/operations/execute_operation_v2.js]: I don't think we get much value out of combining these. I'd prefer splitting into two checks, and throwing informative errors for both cases

File: lib/operations/common_functions.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+function defineAspects(operation, aspects) {
```
Comment (mbroadst) [inline lib/operations/common_functions.js]: nit: aspects are defined in `OperationBase`, should `defineAspects` live there as well?

**Pull request merged**